add option to pass in pre-computed signatures

rchan26 opened the pull request:
- Allow passing on pre-computed signatures into `SignatureMahalanobisKNN` rather than having to compute them
  - Signatures are only computed if `X` is provided and `signatures` is not in the `.fit()` method (similarly for `conformance`)
- Added docstrings for #1 (change to numpy docstring format)
- Added type hints for #2

rchan26 committed:
add option to pass in pre-computed signatures

rchan26 committed:
add default args for signature transform

rchan26 committed:
add type to mahal distance so arrays are all the same type when computing

rchan26 committed:
Merge branch 'main' into pre-comp-signature-passing

rchan26 committed:
Merge branch 'main' into pre-comp-signature-passing

rchan26 committed:
Merge branch 'main' into pre-comp-signature-passing

sz85512678 commented on src/signature_mahalanobis_knn/sig_mahal_knn.py:
```diff
@@ -5,117 +5,179 @@
+                raise ValueError(msg)
+            if self.signature_transform is None:
+                msg = "Must fit the model first"
+                raise ValueError(msg)
```
There are code duplications between the preceding 10 lines and line 63-73 (approximately), is it possible to write it as a separate logic for better maintainability and readability?

rchan26 replied:
sure, will do

sz85512678 replied:
That would be great thanks.

rchan26 replied:
done. but it turned out the only repeated code was the msg I think - I've just made this a global variable

sz85512678 commented on src/signature_mahalanobis_knn/sig_mahal_knn.py:
```diff
@@ -124,7 +186,69 @@ def compute_auc_given_dists(self, distances_in, distances_out):
+
+        return roc_auc
+
+    def compute_auc(
```
Why do we need a separate compute_auc function given that there is sklearn built-in roc_auc_score?

rchan26 replied:
this is just from your code as this already existed. we compute the conformance scores in this function and pass it

happy to merge these two methods together if you'd like

sz85512678 replied:
I see, I suggest not to put this inside the class of SigMahaKNN as its relationship with the core function of the class (namely, computing outliers score for time-series) is rather weak, and I think there is limited circumstances where a user would like to use this functionality. But this is not blocking.

sz85512678 commented on src/signature_mahalanobis_knn/sig_mahal_knn.py:
```diff
@@ -5,117 +5,179 @@
-                    "post": Rescals the output signature by multiplying the depth-d term by d!.
-                            Aim is that every term become ~O(1).
-        :param sig_tfm: One of: ['signature', 'logsignature']).
-        :param depth: int, Signature truncation depth.
```
I would prefer to keep these descriptions somewhere, maybe in the documentation, as they are not obvious (even for me.... not to mention the user).

rchan26 replied:
it's a bit non-standard to repeat another codebase's documentation, but happy to add them back. will do this tomorrow

rchan26 replied:
added them back in latest commit

sz85512678 commented on src/signature_mahalanobis_knn/sig_mahal_knn.py:
```diff
@@ -69,6 +69,19 @@ def fit(
                 SignatureTransformer,
             )
 
+            # set default kwargs for signature transformer if not provided
```
Maybe squash commit together.

rchan26 replied:
what does this mean sorry?

sz85512678 commented on src/signature_mahalanobis_knn/sig_mahal_knn.py:
```diff
@@ -24,6 +24,7 @@ def __init__(
         """
         self.signature_transform = None
         self.n_jobs = n_jobs
+        self.mahal_distance = None
```
Do we need the SigMahaKNN object to hold the distance object? What are the situations that we need to use it directly outside of the class?

rchan26 replied:
With Jonathan, we added this in for debugging mainly, it was very useful to have it and you to use the distance object directly and to compute a distance rather than just the KNN to the corpus.

rchan26 replied:
Another reason is that for pynndescent to work, we can only use abstract methods (we have to use the jitted `Mahalanobis.calc_distance` rather than `Mahalanobis.distance`) as it doesn't like having `self` as an argument. There may be other ways around this.

But in short, we need access to the attributes in the distance object to make pynndescent work.

sz85512678 commented:
Good work Ryan, thank you! Minor concerns about coding styles, let me know if you think what I said make sense... maybe not as it's late on Friday!

rchan26 committed:
make pr changes from comments

rchan26 commented:
Latest changes:
- Made the the error messages for not passing `X` or `signatures` or for not fitting the model yet global variables so you can update the error messages in one place (no duplication)
- Added back the documentation for sktime `SignatureTransformer` arguments and info for what we set as default arguments
- Moved the `compute_auc` methods in `SignatureMahalanobisKNN` to standalone functions outside the class in `utils.py`

rchan26 committed:
add model not fitted global var

sz85512678 commented:
Thanks Ryan for the changes :).
